### ISSUE TITLE
[Feature Request] disable key-chain

### ISSUE THREAD

**Issue by robsdedude** [jkraemer/mail-gpg#55]
Hi,

in my app I only want to use keys. So I do not want to look up keys given a mail address. If the key (give as string) does not work I would like an exception to be raised.

I think to implement this one would need to warp [this](https://github.com/jkraemer/mail-gpg/blob/master/lib/mail/gpg/gpgme_helper.rb#L126) line in a options check and raise some error, if the key couldn't be imported. It might well be that there are other things that have to be changed for this.

**Comment by duckdalbe** [comment]
In the line you reference, `k` is the fingerprint(s) from the imported key-material you provided. If that lookup results in an empty set, [GpgmeHelper#17](https://github.com/jkraemer/mail-gpg/blob/cbe381f9d982ce92bee960a5b4dc423be2addcee/lib/mail/gpg/gpgme_helper.rb#L17) raises an exception.

To me that looks like just what you want —  right?

**Comment by robsdedude** [comment]
Nope. If I call `GPGME::Key.import(k).imports.map(&:fpr)` with an invalid key (pub-key as string) I get an empty Array. Then
https://github.com/jkraemer/mail-gpg/blob/cbe381f9d982ce92bee960a5b4dc423be2addcee/lib/mail/gpg/gpgme_helper.rb#L126

Will look up the recipient as fallback.

**Comment by duckdalbe** [comment]
As far as I understand, the code of GpgmeHelper doesn't work as you described, because

```ruby
[] || 'something'
=> []
```

But the code actually does behave similar to your description (and not like I described), because
`GPGME::Key.find()` returns _all_ known keys if given an empty array:

```ruby
GPGME::Key.find(:public, [], :encrypt)
=> [#<GPGME::Key pub ...]
```

So apparently a new option is required to stop the code-flow if no (usable) key could be imported from the given material.

It would also make the code's behaviour less surprising, in my eyes. So I second this request.

**Comment by robsdedude** [comment]
Sorry I'm actually a Pythonist and thought 

    [] || 'something'
    => 'something'

But anyway. Just as you said I just noticed that (only on some systems) `mail-gpg` encrypts a mail with all pub-keys in the key ring if given an invalid key (#31)

**Comment by jkraemer** [comment]
Does that commit above (which is also part of just released v0.3.2) solve your problem or would you still require an explicit option?

**Comment by jkraemer** [comment]
I think that might not have been enough yet to really raise an error in case of no keys. please try it out with master to include the compact call I just added.

**Comment by robsdedude** [comment]
Ok, sorry for taking so long to give you feedback. When given an invalid key `Mail::Gpg::GpgmeHelper.keys_for_data([mail_address], {mail_address => brokenkey})` will lookup the mail address in the key chain and return only that key. That's an improvement over returning all keys in the key chain but as said, I'd expect an error to be raised here.

Note: The result will be an empty array when the mail address is not stored in the key chain.

**Comment by jkraemer** [comment]
Still no error but the just-released version 0.4 should not use any keys you did not specify in the `:keys` option.

Regarding the option to raise an error, just to make sure I get this right: you want an error to be raised whenever there is a recipient for whom no valid key was present in the :keys argument, correct?

**Comment by robsdedude** [comment]
That would be one option (like `force_key_import: true` or so). The other way (which I thought of in the first place) is to raise the error, when there is a key for a recipient specified for in `:keys` but it cannot be imported.

Like so:
`mail to: [mail_a, mail_b], gpg: {encrypt: true, keys: {mail_a => valid_key}}` would use that `valid_key` for `mail_a` and look up the key for `mail_b`.
However, `mail to: [mail_a, mail_b], gpg: {encrypt: true, keys: {mail_a => invalid_key}, validate_key: true}` (I just made up that option and it's name) would still import the key for `mail_b` but raise an exception as soon as it tries to import the `invalid_key` and fails in doing so. I even think, that this behavior should be the default. Silently using fallbacks when the parameters are invalid isn't good programming style I guess. So probably there should be an option `keychain_fallback` or so which makes gem behave like it does now.
